### PR TITLE
Using prodMode configuration

### PR DESCRIPTION
**Pull request by jessiejuachon**

### CODE REVIEW

File: src/main/java/com/vmware/vipclient/i18n/VIPCfg.java [hunk 1]
```diff
@@ -186,6 +186,8 @@ public void initialize(String cfg) throws VIPClientInitException {
         }
         if (prop.containsKey("sourceLocale"))
         	LocaleUtility.setSourceLocale(Locale.forLanguageTag(prop.getString("sourceLocale")));
+        if (prop.containsKey("prodMode"))
+        	this.prodMode = true;
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/VIPCfg.java:190]: The value may be false.

File: src/main/java/com/vmware/vipclient/i18n/common/ConstantsMsg.java [hunk 2]
```diff
@@ -13,5 +13,6 @@
     public static final String SERVER_CONTENT_ERROR = "The content from server is wrong!";
     public static final String UNKNOWN_ERROR        = "Unknown error.";
     public static final String GET_MESSAGES_FAILED  = "Failed to get messages for component {0}, locale: {1}, data source: {2}.";
+    public static final String GET_MESSAGE_FAILED	= "Failed to get any message for key: {0} of component {1}.";
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/common/ConstantsMsg.java:16]: Should add locale?

File: src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java [hunk 3]
```diff
@@ -36,19 +36,21 @@ public String getString(MessagesDTO dto) {
     	Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java:36]: This is only used in the if block below.

File: sample-client-app/src/main/resources/sampleconfig.properties [hunk 4]
```diff
@@ -1,27 +1,28 @@
...
+offlineResourcesBaseUrl = offlineBundles
+initializeCache = false
+pseudo = false
+collectSource = false
+cleanCache = false
+cacheExpiredTime = 8600000
+machineTranslation = false
+defaultLocale = fr
+sourceLocale = en
+prodMode = true
```
Comment (Xiaochao8) [inline sample-client-app/src/main/resources/sampleconfig.properties:28]: This should be false by default.

File: src/main/resources/vipconfig.properties [hunk 5]
```diff
@@ -22,4 +22,5 @@ collectSource = false
 cleanCache = false
 cacheExpiredTime = 8600000
 machineTranslation = false
-defaultLocale = en
\ No newline at end of file
+defaultLocale = en
+prodMode = true
```
Comment (Xiaochao8) [inline src/main/resources/vipconfig.properties:26]: This should be false by default.

File: sample-client-app/src/main/resources/sampleconfig.properties (hunk 4 above)
Comment (jessiejuachon) [reply sample-client-app/src/main/resources/sampleconfig.properties:28]: It is false by default in the client library. This is just a sample.

File: src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java (hunk 3 above)
Comment (jessiejuachon) [reply src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java:36]: It is used twice: in lines 41 and 42.

File: src/main/resources/vipconfig.properties (hunk 5 above)
Comment (jessiejuachon) [reply src/main/resources/vipconfig.properties:26]: It is false by default in the code. This is just for testing purposes.

File: sample-client-app/src/main/resources/sampleconfig.properties (hunk 4 above)
Comment (jessiejuachon) [reply sample-client-app/src/main/resources/sampleconfig.properties:28]: I will just manually set prodMode=true in one of the examples.

File: src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java [hunk 6]
```diff
@@ -36,19 +36,21 @@ public String getString(MessagesDTO dto) {
...
-    		Locale fallback = fallbackLocalesIter.next();
-			MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), 
-					key, dto.getSource(), fallback.toLanguageTag(), null);
-			cacheItem = new ComponentService(fallbackLocaleDTO).getMessages();
-			
-			// Cache a reference to the MessageCacheItem of the fallback locale 
-			if (!cacheItem.getCachedData().isEmpty()) {
-				CacheService cacheService = new CacheService(dto);
-				cacheService.addCacheOfComponent(cacheItem);
+    	if (VIPCfg.getInstance().isProdMode()) {
```
Comment (jessiejuachon) [inline src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java:39]: If statement added for use case #1: Requested locale is not supported (bundle not found onlline or offline)

File: src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java [hunk 7]
```diff
@@ -96,7 +97,9 @@ public String getMessage(final Locale locale, final String component, final Stri
     		if (source != null && !source.isEmpty()) {
     			return FormatUtils.format(source, LocaleUtility.getSourceLocale(), args);
     		}
-    		return key;
+    		if (VIPCfg.getInstance().isProdMode())
```
Comment (jessiejuachon) [inline src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java:100]: If-statement added for use case#2: A requested (target/default) message is not found; source message is also not found.

File: src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java [hunk 8]
```diff
@@ -63,8 +67,8 @@ public TranslationMessage() {
      */
     private String getCachedMessage(String component, String key, Locale locale) {
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java:68]: How about to change name to getMessageFromCache?

File: src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java [hunk 9]
```diff
@@ -305,17 +306,38 @@ public boolean postString(final Locale locale, final String component,
...
+    	while (cacheItem.getCachedData().isEmpty() && fallbackLocalesIter.hasNext()) {
+    		Locale fallback = fallbackLocalesIter.next();
+			MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), 
+					dto.getKey(), dto.getSource(), fallback.toLanguageTag(), null);
+			cacheItem = new ComponentService(fallbackLocaleDTO).getMessages();
+			
+			// Cache a reference to the MessageCacheItem of the fallback locale 
+			if (!cacheItem.getCachedData().isEmpty()) {
+				CacheService cacheService = new CacheService(dto);
+				cacheService.addCacheOfComponent(cacheItem);
```
Comment (Xiaochao8) [inline src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java:336]: The loop should break here.

File: src/main/java/com/vmware/vipclient/i18n/VIPCfg.java [hunk 10]
```diff
@@ -67,7 +67,6 @@
     private String                     i18nScope     = "numbers,dates,currencies,plurals,measurements";
     private String					   offlineResourcesBaseUrl;
     
-
```
[the remark this answers is not in the document]
Comment (jessiejuachon) [reply src/main/java/com/vmware/vipclient/i18n/VIPCfg.java:70]: Why do we need 2 blank lines? I just cleaned this up  a bit.

File: sample-client-app/src/main/resources/sampleconfig.properties [hunk 11]
```diff
@@ -1,27 +1,27 @@
...
-initializeCache = false
-pseudo = false
-collectSource = false
-cleanCache = false
-cacheExpiredTime = 8600000
-machineTranslation = false
-defaultLocale = fr
-sourceLocale = en
\ No newline at end of file
+# It's used for categorizing translation by product on Singleton side to maintain translation.
```
[the remark this answers is not in the document]
Comment (jessiejuachon) [reply sample-client-app/src/main/resources/sampleconfig.properties:1]: It had carriage returns at the end. I just cleaned it up.
https://stackoverflow.com/questions/1822849/what-are-these-ms-that-keep-showing-up-in-my-files-in-emacs

File: src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java (hunk 9 above)
Comment (jessiejuachon) [reply src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java:336]: If !cacheItem.getCachedData().isEmpty(), then the while loop condition is not satisfied anymore in the next loop. So it naturally breaks, no need to add a break statement.

File: src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java [hunk 12]
```diff
@@ -26,17 +26,19 @@
...
     @SuppressWarnings("unchecked")
+    @Deprecated
     public String getString(MessagesDTO dto) {
     	String key = dto.getKey();
     	
-    	Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
     	MessageCacheItem cacheItem = new ComponentService(dto).getMessages();
     	
     	// While failed to get MessageCacheItem, use MessageCacheItem of the next fallback locale. 
+    	Iterator<Locale> fallbackLocalesIter = LocaleUtility.getFallbackLocales().iterator();
```
[the remark this answers is not in the document]
Comment (jessiejuachon) [reply src/main/java/com/vmware/vipclient/i18n/messages/service/StringService.java:41]: Do you mean line 41? I just moved it from line 36 to here, next to the while loop where it is used.

File: src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java [hunk 13]
```diff
@@ -305,17 +306,38 @@ public boolean postString(final Locale locale, final String component,
...
+			}
```
Comment (jin-jasper) [inline src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java:337]: 329~338 has indention problem.

File: src/main/java/com/vmware/vipclient/i18n/common/ConstantsMsg.java [hunk 14]
```diff
@@ -13,5 +13,6 @@
     public static final String SERVER_CONTENT_ERROR = "The content from server is wrong!";
     public static final String UNKNOWN_ERROR        = "Unknown error.";
     public static final String GET_MESSAGES_FAILED  = "Failed to get messages for component {0}, locale: {1}, data source: {2}.";
+    public static final String GET_MESSAGE_FAILED	= "Failed to get any message for key: {0} of component {1}, requested locale: {2}";
```
Comment (jin-jasper) [inline src/main/java/com/vmware/vipclient/i18n/common/ConstantsMsg.java:16]: alignment problem of equal sign

File: src/test/java/com/vmware/vip/i18n/TranslationMessageTest.java [hunk 15]
```diff
@@ -211,7 +220,7 @@ public void testGetComponentMessages() {
         Assert.assertEquals(message_zh_CN, retMap5.get(key));
 
         Map<String, String> retMap6 = translation.getMessages(Locale.forLanguageTag("zh-Hant-TW"), component);
-        Assert.assertEquals(message_zh_TW, retMap6.get(key));
+        Assert.assertEquals(message_zh_TW, retMap6.get(key));
```
Comment (jin-jasper) [inline src/test/java/com/vmware/vip/i18n/TranslationMessageTest.java:223]: There is a space char or a tab at the end of the line.

File: src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java (hunk 9 above)
Comment (Xiaochao8) [reply src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java:336]: Don't need to judge twice in one cycle.

File: src/test/java/com/vmware/vipclient/i18n/messages/service/OfflineModeTest.java [hunk 16]
```diff
@@ -106,16 +110,19 @@ public void testGetMsgsFailedKeyNotFound() {
...
+    	// Throw an exception because message key does not exist anywhere
+    	assertEquals(FormatUtils.format(ConstantsMsg.GET_MESSAGE_FAILED, key, component, newLocale), e.getMessage());
     	
     	cfg.setOfflineResourcesBaseUrl(offlineResourcesBaseUrlOrig);
     	cfg.setMsgOriginsQueue(msgOriginsQueueOrig);
     }
     
     @Test
-    public void testGetMsgsFailedUseDefault() { 
+    public void testGetMsgsFailedProdModeUseDefault() {
```
Comment (jin-jasper) [inline src/test/java/com/vmware/vipclient/i18n/messages/service/OfflineModeTest.java:125]: 'ProdMode' should be deleted.

File: src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java (hunk 9 above)
Comment (jessiejuachon) [reply src/main/java/com/vmware/vipclient/i18n/base/instances/TranslationMessage.java:336]: I don't get it. What do you mean by judge twice in one cycle?

**Pull request merged**